### PR TITLE
Change SystemConfig to use SystemConfigKey instead of constants

### PR DESCRIPTION
**Pull request by billonahill**
Refactor to use enum keys for `SystemConfig` class. Introduce a builder so `SystemConfig` can become immutable, and refactor `Simulator` and test classes to use it for custom settings.

Will follow up with refactoring other consumers to use builder and get rid of the use of `public SystemConfig(String configFile, boolean mustExist)` constructor.

### CODE REVIEW

**Comment by billonahill** [comment]
I went ahead and removed usage of `public SystemConfig(String configFile, boolean mustExist)` constructor in favor of the Builder approach since it wasn't that big of a change.

File: heron/common/src/java/com/twitter/heron/common/config/SystemConfigKey.java [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+import java.util.HashSet;
+import java.util.Set;
+
+import com.twitter.heron.common.basics.ByteAmount;
+
+/**
+ * Keys to be used in the SystemConfigs
+ */
+// TODO(mfu): Provide default values for all configs uniformly
+// TODO(mfu): https://github.com/twitter/heron/issues/970
```
Comment (objmagic) [inline heron/common/src/java/com/twitter/heron/common/config/SystemConfigKey.java]: is this TODO addressed?

**Comment by nlu90** [comment]
👍

File: heron/common/src/java/com/twitter/heron/common/config/SystemConfigKey.java (hunk 1 above)
Comment (billonahill) [inline heron/common/src/java/com/twitter/heron/common/config/SystemConfigKey.java]: We now have a mechanism for defaults and we use it for 2 config keys. @maosongfu do you think that's enough for #970 or should we also be adding defaults for other key values. You comment was initially above keys that now have a default, so maybe we're ok now?

**Pull request merged**